tests/client/net_handler_client.c

```diff
@@ -0,0 +1,520 @@
+	while (1) {
+		//set up the read_set argument to selct()
+		FD_ZERO(&read_set);
+		FD_SET(nh_tcp_shep_fd, &read_set);
+		FD_SET(nh_tcp_dawn_fd, &read_set);
+		FD_SET(nh_udp_fd, &read_set);
+		
+		//prepare to accept cancellation requests over the select
+		pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
+		
+		//wait for something to happen
+		if (select(maxfd, &read_set, NULL, NULL, NULL) < 0) {
+			printf("select: output dump: %s\n", strerror(errno));
+		}
+		
+		//deny all cancellation requests until the next loop
+		pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
+		
+		//enable tcp output if more than enable_thresh has passed between last time and previous time
+		if (FD_ISSET(nh_tcp_shep_fd, &read_set) || FD_ISSET(nh_tcp_dawn_fd, &read_set)) {
+			curr_time = millis();
+			if (curr_time - last_received_time >= enable_threshold) {
+				less_than_disable_thresh = 0;
+				tcp_output_fp = stdout;
+			}
+			if (curr_time - last_received_time <= disable_threshold) {
+				less_than_disable_thresh++;
+				if (less_than_disable_thresh == sample_size) {
+					printf("Suppressing output: too many messages...\n\n");
+					fflush(stdout);
+					tcp_output_fp = null_fp;
+				}
+			}
+			last_received_time = curr_time;
+		}
+		
+		//if we were asked to print the next UDP, set the UDP pointer
+		pthread_mutex_lock(&print_udp_mutex);
+		if (print_next_udp) {
+			udp_output_fp = stdout;
+		}
+		pthread_mutex_unlock(&print_udp_mutex);
+
+		//print stuff from whichever file descriptors are ready for reading...
+		if (FD_ISSET(nh_tcp_shep_fd, &read_set)) {
+			if (recv_tcp_data(SHEPHERD, nh_tcp_shep_fd) == -1) {
+				return NULL;
+			}
+		}
+		if (FD_ISSET(nh_tcp_dawn_fd, &read_set)) {
+			if (recv_tcp_data(DAWN, nh_tcp_dawn_fd) == -1) {
+				return NULL;
+			}
+		}
+		if (FD_ISSET(nh_udp_fd, &read_set)) {
+			recv_udp_data(nh_udp_fd);
+		}
+	}
+	return NULL;
+}
+
+// ************************************* NET HANDLER CLIENT FUNCTIONS ************************** //
+
+void start_net_handler ()
+{
+
+	//fork net_handler process
+	if ((nh_pid = fork()) < 0) {
+		printf("fork: %s\n", strerror(errno));
+	} else if (nh_pid == 0) { //child
+		//cd to the net_handler directory
+		if (chdir("../../net_handler") == -1) {
+			printf("chdir: %s\n", strerror(errno));
+		}
+
+		//exec the actual net_handler process
+		if (execlp("./net_handler", "net_handler", (char *) 0) < 0) {
+			printf("execlp: %s\n", strerror(errno));
+		}
+	} else { //parent
+		sleep(1); //allows net_handler to set itself up
+
+		//Connect to the raspi networking ports to catch network output
+		nh_tcp_dawn_fd = connect_tcp(DAWN_CLIENT);
+		nh_tcp_shep_fd = connect_tcp(SHEPHERD_CLIENT);
+		if ((nh_udp_fd = socket(AF_INET, SOCK_DGRAM, 0)) == -1) {
+			printf("socket: UDP socket creation failed...\n");
+			stop_net_handler();
+			exit(1);
+		}
+		memset(&udp_servaddr, 0, sizeof(struct sockaddr_in));
```

Comment:
👌

Reply:
Ok, actually I can't do this here, since `udp_servaddr` is a global variable. I have to use `memset`. Is there a reason why you prefer the 0 struct initializer over `memset`?

Reply:
Isn't this variable only used once, or am I mistaken? Because if its used once you can initialize it to 0. Talked about it on slack but I prefer the 0 initializer since its much cleaner.